sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -115,12 +115,6 @@ private[hive] class HiveClientImpl(
   private val outputBuffer = new CircularBuffer()
 
   private val shim = version match {
-    case hive.v12 => new Shim_v0_12()
-    case hive.v13 => new Shim_v0_13()
-    case hive.v14 => new Shim_v0_14()
-    case hive.v1_0 => new Shim_v1_0()
-    case hive.v1_1 => new Shim_v1_1()
-    case hive.v1_2 => new Shim_v1_2()
     case hive.v2_0 => new Shim_v2_0()
```

Comment:
seems the above `Shim_v*` classes could be removed and merged the code into `Shim_v2_0`

Reply:
I think this is really fine to keep them. I made them all private for now, and explicitly throw an exception if other versions are specified.

Reply:
This inherits each other, and it's different from `v14`, etc.

Reply:
Emm.. what's the benefit of retaining unreachable code?

For example, all of `Shim_v0_12`, `Shim_v0_14` and `Shim_v2_0` define `loadPartition`, after this patch, the `loadPartition` defined in `Shim_v0_12`, `Shim_v0_14` becomes dead code.

Reply:
Those dead code isn't really needed but keeping them as are doesn't risk, and they are well organized according to versions.

I also worry about downstram users/vendors who still need to use them, and would like to alleviate the overhead.

Maybe let's do this in a separate PR.

Reply:
> I also worry about downstram users/vendors who still need to use them, and would like to alleviate the overhead.

I think they could easily find them in the old versions or from the git history.

> Maybe let's do this in a separate PR.

I'd like to make such a PR if you don't oppose

Reply:
> I'd like to make such a PR if you don't oppose
Yes please :-).